### PR TITLE
Add `ScalarMix` module

### PR DESCRIPTION
**Pull request by matt-peters**
Add `ScalarMix` module that computes scalar weighted averages of tensors, with optional layer normalization

### CODE REVIEW

File: allennlp/modules/scalar_mix.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        super(ScalarMix, self).__init__()
+
+        self.mixture_size = mixture_size
+        self.do_layer_norm = do_layer_norm
+
+        self.scalar_parameters = ParameterList([Parameter(torch.FloatTensor([0.0]))
+                                                for _ in range(mixture_size)])
+        self.gamma = Parameter(torch.FloatTensor([1.0]))
+
+    def forward(self, tensors: List[torch.Tensor],  # pylint: disable=arguments-differ
```
Comment (DeNeutoy) [inline allennlp/modules/scalar_mix.py]: Please can you add a tiny docstring here and explain that the mask is for all of the tensors that are being mixed.

File: allennlp/modules/scalar_mix.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                mask: torch.Tensor = None) -> torch.Tensor:
+        """
+        Compute a weighted average of the ``tensors``.  The input tensors an be any shape
```
Comment (DeNeutoy) [inline allennlp/modules/scalar_mix.py]: @matt-peters Was it really too much effort to match the docstring style of the rest of `allennlp`? I know you think it doesn't matter, but it _does_ make it easier to read and more uniform, which is really the key thing. It's just going to make reviews take longer if we can't approve stuff with changes and expect you to add it in some uniform style.

Comment (matt-gardner) [reply allennlp/modules/scalar_mix.py]: In fairness to @matt-peters, we don't have a place to look for what that style should be, and we haven't always been uniform ourselves.  I'll add a section to the style guide PR I'm putting together that gives guidance on what things should go where for methods like this.

Comment (matt-peters) [reply allennlp/modules/scalar_mix.py]: Yes if you provide a template for a docstring then I'll fill it out.

**Pull request merged**